[#20] 리팩토링 (Item 도메인 제외)

dding94 opened the pull request.

dding94 committed:
refactor: ErrorCode 패키지 위치 변경 response -> exception

나머지 영향 받는 예외 코드 import 수정

dding94 committed:
refactor: ErrorCode 패키지 변경되어 import 수정

dding94 committed:
refactor: ErrorCode 패키지 변경되어 import 수정

dding94 committed:
refactor: 인증, 인가관련 부분 로직 수정

로그인 체크 어노테이션 로그인 정보가 있어야함,토큰 프로퍼티 위치 변경, 필터는 사용하지 않으므로 삭제

dding94 committed:
refactor: spring 의존성 최소화

dding94 committed:
refactor: Order 도메인 리팩토링

빈은 수동 등록 하도록 변경, JpaConfig 빈 이름 충돌 나서 앞에 도메인 이름 붙임

dding94 committed:
refactor: Jpa 설정 정보 추가 및 config 이름 변경

UserProcessorConfig - > UserConfig 이름 변경

dding94 committed:
refactor: get 메서드 변경

dding94 committed:
test: user, order 도메인 테스트 코드 작성

dding94 committed:
refactor: user 도메인 리팩토

application layer 스프링 의존성 분리, adapter 도 따로 Config 로 관리하도록,
Processor 클래스 이름 변경

dding94 committed:
Merge branch 'main' into feature/21-jpa-and-refactor

2eve2 commented on user-api/user-application/src/main/java/com/flab/user/application/facade/UserManager.java:
```diff
@@ -1,44 +1,42 @@
 import com.flab.user.domain.exception.DuplicatedEmailException;
-import org.springframework.stereotype.Service;
 
-@Service
```
spring 의존성을 줄이고자 @Service을 사용하지 않고 @Bean 으로 다 등록되게끔 구현하실 생각이신가요?

무분별한 @Getter @Setter 등의 어노테이션 사용을 지양하는 취지는 얘기가 되었었는데, @Service, @Repository 등까지 사용하지 않는 이유는 잘 모르겠는데 설명해주실 수 있나요? @Configuration도 결국 spring에 종속되어있는거라 생각해서요. 다른 config 클래스들도 많아서 Repository와 Service를 등록하기 위한 config 클래스들도 섞여 config 클래스들이 너무 많아지고 꼭 따로 등록할 필요성이 있는지 궁금합니다.

dding94 replied:
애플리케이션 레이어도 마찬가지로 외부의 특정 기술에 종속적이지 않아야 한다고 생각합니다.
이 계층의 얇게 유지되어야 하고, 작업을 조정하고, 위임하는 역할을 해야하기 때문에 최소화로 했습니다.
그런데 인프라레이어에서는 기술에 직접적으로 구현할 수 있는 계층이니 해당 계층에서는 스프링에 의존적인 코드를 사용해도 무방하다고 생각합니다. 
관리적인 측면에서 수동으로 빈으로 등록하게 만들면 좋겠다고 생각을 했습니다.
클래스가 많아지는것도 사실 계층 2개만 추가되었고 srp를 위배하지 않았기 때문에, 많아져서 문제가 된다고 생각하지는 않습니다.
그래서 gradle 에 의존성을보면 스프링 트랜잭션만 의존해서 사용하도록 만들었습니다!

추가로 무분별한 @getter @setter 의 사용을 지양하는 취지로 얘기드린게 아니라 도메인은 순수하게 pojo 로 구현하고자 편의를 제공하는 롬복도 제외했습니다. 왜냐하면 롬복을 사용하는것도 결국 Dependency 설정을 해야하기 때문이라고 생각했습니다.

제가 잘못된 정보, 생각을 얘기 하고 있다면 피드백 주시면 감사하겠습니다.

2eve2 replied:
아 제가 @getter, @setter 와 @Service, @Repository 등을 구분지어 말씀드린 것은 @getter, @setter와 같은 롬복 어노테이션  제외하는 것에 동의하나 @Service, 등은 롬복이 아닌 org.springframework.stereotype에 해당하므로 롬복 제외랑 관련이 없는 것 같아서 말씀드린거였는데 설명이 짧았나봅니다. 

infra 위치로 종속성이 보여지게해서 클래스를 완전히 pojo 형태로 보여질 수 있게끔 하는 것은 좋은 점 같습니다. 하지만 이미 Manager와 Repository를 등록하는 @Configuration이 스프링에 종속하는데 Manager와 Repository가 스프링에 종속되지 않는다고 할 수 있을까요? 실제로 application, domain이 스프링에 종속된건 똑같은데 보이는 위치만 이동한 것 같아서요. 이 점은 잘 모르겠네요.

저도 "비즈니스 로직을 구현하는 코드를 작성할 때" application 계층이 얇게 유지되어야 하고 도메인에 행동의 책임을 위임하는 것에는 동의합니다. 

관리적인 측면에서 더 좋은 점은 무엇인가요? 기존에 @Service, @Repository 설정으로 컴포넌트 관리하는 것이 @Configuration에 Bean으로 등록하는 것보다 관리적인 측면에서 더 안좋아서 바꿔야하는지는 모르겠습니다. 오히려 Manager와 RepositoryAdapter 등을 보면 바로 어떻게 구성되어있고 의존성을 알 수 있는 상황에서, config로 빈 등록 파트만 따로 분리해서 한번 더 보여주는 것 뿐이라 제 생각으로는 불필요하지 않나 생각됩니다.

그리고 제가 추가로 @Component vs @Configuration을 통한 @Bean 선택에 대한 스택오버플로우와 관련 자료를 찾아보았는데 현재 상황에서는 @Component (@Service, @Repository 등)으로 사용하는 것이 추천되고 있네요.

밑에 관련 자료입니다.

![image](https://user-images.githubusercontent.com/58822617/203797140-a420f360-7e5d-4837-9d6d-a33d71d882c5.png)

https://www.javahabit.com/2018/12/30/spring-bean-vs-component/

dding94 replied:
해당 레이어가 가진 gradle 의 의존성을 보면 어디에 의존(종속)하는지 알 수 있습니다.
![스크린샷 2022-11-24 오후 10 40 43](https://user-images.githubusercontent.com/21376853/203798249-180e31c6-d259-487e-8526-a87440344802.png)
위 사진은 애플리 케이션 레이어가 의존하는 방향입니다.

![스크린샷 2022-11-24 오후 10 41 12](https://user-images.githubusercontent.com/21376853/203798351-6f55b81b-fd84-4b0b-a456-c33a80c1582b.png)
위 사진은 인프라 레이어가 의존하는 방향입니다.

gradle 에 셋팅한 의존성 방향을 보면 애플리케이션레이어가 스프링 트랜잭션, 도메인을 제외하고 의존성을 가지지 않으니 스프링 프레임워크에 종속적이지 않다고 할 수 있지않을까요? 도메인 또한 그렇구요 도메인계층도 정말 스프링에 종속됐다고 말할 수 있을까요
![스크린샷 2022-11-24 오후 11 01 18](https://user-images.githubusercontent.com/21376853/203802610-f0c3a989-8974-4cb0-929a-ca5c88d0f547.png)
위 사진은 도메인계층이 의존하는 것들입니다. (common 도 직접 보시면 의존성 설정이 없고, 순수한 자바로 구성되어있습니다)

현재 인프라 레이어가 모든 계층을 의존하고 있으니, 의존성 방향에 대해서 반대로 생각하셔서 애플리케이션, 도메인이 스프링에 종속적이라고 얘기하신게 아닐까라는 생각이 듭니다.

애플리케이션 레이어는 도메인 객체들을 가져와서 조정을 통해서 유스케이스를 만족시키는 기능을 제공하는 책임을 가지고 있다고 봅니다. 그러면 특별한 기술 의존성이 필요 없는 계층이 아닌가 생각이 듭니다.

관리적인 측면인 인프라 계층의 config 패키지에서 어떤 설정이 되어있는지 한눈에 확인할 수 있기 때문입니다.

마지막 부분만 보면 뭔가 저희 프로젝트를 전부 component 로 해야하는게 아닌가 생각이 들어서, 괜찮으시면 좀 더 설명을 더 해주시면 감사하겠습니다!

2eve2 replied:
말씀해주신대로 build.gradle로 dependency를 차이를 확인해보고자 UserManager쪽에 @Service를 추가하고 UserFacadeConfig 파일을 지운 뒤 프로젝트 실행과 build.gradle 확인을 해보았습니다. 

그런데 @Service를 지웠을 때와 안지웠을 때 build.gradle 차이가 없이 잘 실행이 됩니다. 도메인, configuration, jpa 등만 의존하고 있습니다. 즉 @Service 를 붙이든 안붙이든 스프링 의존성 차이가 없이 보입니다. 한번 확인해주시고 혹시 제가 잘못한 부분이 있다면 알려주시면 감사하겠습니다. 
![image](https://user-images.githubusercontent.com/58822617/203878439-68f452a2-38f3-4b1e-a84b-97a7413bef92.png)


그리고 스프링 의존성을 infra layer에서만 주어야한다는 가정으로 @Repository에서 이해가 안가는 것이 있는데요. 기존에도 ~RepositoryAdapter라는 infra layer에서 @Repository 선언을 해주므로 도메인에 Repository 클래스에 스프링 의존성이 없도록 설계했었지 않나요? 


또한 마지막 부분은 제가 이해하기론 third party 기술이나 스프링을 사용하지 않고 개발자가 자유로이 제어가 불가능한 영역이 아니면 component로 사용하는 것이 좋다는 글로 이해했습니다. User Api를 예로 들면, infra의 config 영역에서 TokenRepositoryConfig와 UserConfig는 구체적으로 RedisTokenRepository와 BcryptPasswordEncoder 및 thirdparty 기술에 의존되어 있고, 비즈니스 로직을 직접적으로 구현하는 영역 특성상 서비스가 확장될 시 이처럼 다른 외부 기술 혹은 개발자가 만든 infra 영역의 구현체에 의존할 가능성이 많습니다. 따라서 @Configuration 을 통한 @Bean 으로 관리하는 것이 합리적인 것 같습니다. 

하지만 UserFacadeConfig와 UserRepositoryConfig는 스프링 프레임워크 내에서 third party 기술과 상관없이 개발자가 작성한 클래스이므로 Component가 더 알맞다고 이해했습니다.

2eve2 replied:
추가로 저 위에 이미지로 제시한 의존성 보면 UserApplicationlayer에서 jpa가 의존되어 있는데 이것도 제외해야할 것 같습니다..?

dding94 replied:
첫번째 하신 방법은 차이가 없어야 합니다. 인프라레이어에서 수동으로 등록한 UserFacadeConfig를 지우시고 애플리케이션레이어에서 service를 사용하여 매니저를 bean 으로 등록 하는 방법을 하신거죠. 이는 component 를 사용한거랑 다름이 없습니다.

추가로 의존성을 체크하기 좋은 방법중 하나인 import 문을 확인하면 되는데 애플리케이션 레이어에서 'service' 를 사용하기 위해 import org.springframework.stereotype.Service 를 선언하여 의존하고 있지 않습니까? 그래서 해당 계층이 스프링에 의존적하고있다 라고 말할 수 있던 것 입니다.

잘못 알고 계신점이 있어서 얘기 드립니다. 현재 애플리케이션 레이어는 'jpa' 가 아니라 'spring-jpa' 에 의존하고 있기 때문에 spring 에서 제공하는 service 를 사용 가능한 상태여서 그렇습니다. 마지막 말씀해주신 것 처럼 spring jpa 의존성을 제외한다면 service 를 사용하지 못합니다.
참고로 jpa 에 의존하려면 implementation 'javax.persistence:javax.persistence-api:x.x' 를 사용해야 합니다.
(해당 도메인은 수정이 필요한 도메인 입니다.  Order 도메인을 중점으로 보시면 될 것 같습니다.)

component 를 붙여도 해당 방법이 틀린건 아니라고 생각합니다. 하지만 제가 위에서 언급한 계층의 특성을 생각해보면 component 를 사용하지 않아도 될 것 같다고 충분한 이유를 설명드린 것 같아서 이부분은 넘어가겠습니다.

2eve2 replied:
아하 제가 jpa와 component에 대해 잘못 알고 있는 부분이 있었던 것 같습니다. 

그래도 말씀드린 요지는 똑같은데요. 
각자 의견을 정리해서 각 layer 별로 저희가 컨벤션을 어떻게 정할건지 논의를 해봐야할 것 같습니다.

우선 제 의견은 멘토님 질문과 위의 리뷰에서도 말씀드렸지만, application layer에서 "스프링 의존성을 없애기 위해" 빈 등록을 분리할 필요는 없다고 생각합니다. 
application layer에서 @Service를 사용하지 않는다고 해서 spring 의존성을 완전히 분리하기에는 어렵기에 말씀하신 계층의 특성을 이유로 따로 Bean 등록을 할 필요는 없다고 생각하기 때문입니다. application layer에서 첨부해주신 이미지처럼 결국 transactional을 사용하기에 spring 의존성을 완전히 분리하기도 어렵고, 이것 또한 분리했을 때 장점을 크게 못느끼겠습니다.

또한 @Repository는 이미 infrastructure 내에서 RepositoryAdapter에서 사용하고 있는데 config에서 한번 더 빈으로 따로 분리할 필요가 있나 싶습니다.

다만 domain layer는 지금처럼 최대한 pure하게 만드는 것이 맞다고 생각합니다.

우선 의존성 관리 측면 외에 계층별 특성을 이유로 말씀하신 부분만 의견을 달았습니다.

2eve2 replied:
제 생각엔 스프링 의존성 측면 말고 관리적인 이점만 생각했을 때에도 이미 ~Config(ex. OrderConfig)내에서 사용되는 processor들이 무엇이 있고, 어떤 의존성을 갖는지 다 보여주고 있어서 크게 필요성을 못느끼겠는데 어떻게 생각하시나요?

dding94 replied:
제가 말을 좀 애매하게 한거 같아서 확실하게 말씀드리면 스프링 의존성을 없애기보다 **의존성을 최소화** 하는게 저의 목표입니다. 
추가로 저희가 user 인프라 계층에서 **시큐리티의 의존성을 최소화** 하기위해 implementation 'org.springframework.security:spring-security-crypto' 사용을 했습니다. 의존성을 완전히 분리시키기 위함이 아니라 최소화 하기 위함이였죠. 저희가 사용하지 않을 시큐리티에 대한 의존성을 전부 걸 필요가 없기 때문입니다. 애플리케이션레이어도 마찬가지로 스프링에 대해서 완전히 의존성을 분리하기 어렵다면, 의존성을 최소화하는게 좋지 않을까 생각이 듭니다.

2eve2 replied:
음..  application layer의 의존성을 최소한으로 하기 위해 FacadeConfig는 분리하되, RepositoryConfig는 분리하지 않는 방향을 제안해봅니다. 말씀하신 의존성 최소화의 목적이라면 기존처럼 infra layer의 RepositoryAdapter에서 @Repository를 관리해도 문제가 없다고 생각하고 필요성이 적은 것 같습니다.

f-lab-bright replied:
여기서 살짝 의견을 하나 남겨두자면, 어떤 부분을 Config로 관리했을 때 이점이 크고, 어떤 부분을 Config로 관리하지 않아야 이점이 큰지 고민해보시면 좋겠습니다.
Facade (지금의 ~~Manager 들) 의 경우에는 대부분 usecase or 가끔 repository 정도를 주입받아서 인터페이스를 하나로 묶어주는 역할을 주로 담당할텐데, 이 facade의 의존성 주입이 바뀌는 일이 얼마나 있을까요?? 혹은 이 facade의 유지보수를 위해서 어떤 의존성이 주입되는지 파악하기 위해서 드는 수고가 얼마나 있을까요??
Repository의 경우에는 어떤 책임을 같는 녀석들에게 주입이 되나요?? 그 때 어떤 녀석을 빈으로 사용할지 결정되는 지점이 어디일까요??

dding94 replied:
저희가 돌아가서 bean을 등록하는 책임을 누가 관리할지 의논을  해봐야 할 것같습니다.
저도 인프라에서는 위에서 언급한 것 처럼 굳이 repository 를 붙여 관리해도 문제는 없다고 생각합니다.
저는 config 라는 클래스가 bean을 등록하는 책임을 가진다고 생각하여 직접 수동으로 하는 것도 적합하다고 생각합니다. 나중에 service 나 repository 가 버전업으로 변경이 되거나 그랬을때 config 에서만 관리하기 때문에 해당 부분만을 고치면 된다고 생각하기 때문입니다.
다정님은 이런 관점에서 생각했을 때 누가 bean 을 등록하는 책임을 가지도록 설계하면 좋다고 생각하시나요?

f-lab-bright replied:
그리고 첨언하자면, spring-security-crypto 를 사용하라고 말씀드렸던건 의존성의 최소화하는 목적이 아니라, 스프링 시큐리티에서 사용하는 암호화 방식은 그대로 사용하되, 스프링 시큐리티의 필터 기능은 사용하지 않고 직접 구현해보라고 권해드린거지 의존성을 최소화하는 관점에서 이걸 사용하라고 가이드를 드린건 아닙니다.
그리고 모듈 단위로 나뉘어졌을 때 해당 모듈에 이미 의존성이 추가되었다면 이미 그 모듈은 해당 의존성에 의존하는 구성이 되어버린거라고 보셔야 합니다.
언제든지 사용할 수 있는 상태가 되었으니까요.
여기서 의존성을 덜 사용하게 하려면 모듈에서의 의존성에서부터 달라져야하지, 코드 컨벤션적으로 의존성을 덜 사용하는건 협업자가 많아질수록 더 무의미해집니다.
모듈 단위로 나누어서 의존성이 강제되는 수준이 되는 정도여야 의미가 있고, 개별 모듈 내에서는 크게 유의미하지 않다고 생각하시면 됩니다.
협업자가 많아질수록, 시간이 지날수록 이 컨벤션은 점점 유지되기 힘들테니까요.
본인이 작업 할 때 외의 사람이 작업하면서 금방 포이즈닝 되는 설계는 사실상 의미가 없다고 생각하시면 되기 때문에 의존성을 최소화하는 등의 무언가를 하시고 싶다면 최대한 강제적으로 그렇게 작성할 수 밖에 없는 구성을 고민해보시는게 유의미합니다 :)
즉 현재의 의존성 구성으로는 application layer에서 @Service 등의 스프링 어노테이션을 사용하지 않아야 할 이유는 크게 없다고 볼 수 있겠네요 :)

2eve2 replied:
어떤 것을 config로 관리하는 것이 이점이 큰지를 생각해보았을 때, Facade에서 의존성 주입이 바뀌는 일은 새로운 기능이 추가되어 processor가 추가로 등록되었을 때 제외하고는 바뀔 일이 적으며 Facade에 어떤 의존성이 주입되는지 파악하기 위해 config를 분리해서 봐야할 정도로 수고스럽지 않다고 생각합니다. 따라서 FacadeConfig의 필요성이 적다고 생각합니다.

Repository는 해당 Repository를 사용하는 application processor에서 주입이 되고, 어떤 녀석이 빈으로 사용할지 결정되는 지점은 잘 모르겠습니다..

f-lab-bright replied:
repository 부분은 실제 구현 부가 바뀌는 일이 생길 수 있을만한 녀석이고 (저희가 mybatis / jdbc에서 jpa로 변경했듯이), usecase에서 실제로 사용될 repository가 어떤 녀석으로 선택될지를 확인할 수 있게 구성하는게 좋지 않을까요?
여러 구현된 repository들이 있을 때 어떤 녀석이 실제로 빈으로 등록되는지, 혹은 어떤 녀석이 실제로 주입되는지를 파악하기 쉬우려면요 :)

dding94 replied:
제가 그때 가이드를 해주신 내용을 잘 못 받아드려 의존성 최소화 관점을 생각 한 것 같습니다.
지난 pr 에서도 애플리케이션이 starter-web 으로 설정해 뒀는데 servlet 에 종속된 모듈은 아니라고 하셔서 이 부분도 마찬가지로 의존성 최소화로 받아 드린 것 같습니다.
궁금한점은 포이즈닝 이라는 말이 막힌다는 설계 이런 것 일까요?

f-lab-bright replied:
잘못된 코드가 금방 전파될 수 있다는 의미입니다.
모듈에 의존성이 추가된 상태라면 우리가 아무리 컨벤션으로 이걸 사용하지 말자라고 정해두더라도, 팀의 인원이 바꾸고, 시간이 지나면서 우리가 의도하지 않은 형태로 코드 설계가 오염될 수 있겠죠.
가능하면 이런 오염을 근본적으로 방지할 수 있는 형태의 코드 (의존성) 설계를 하시는게 좀 더 오래 잘 유지되는 좋은 설계를 유지할 수 있습니다 :)

2eve2 replied:
네 그렇네요 :)

그렇다면 구체적인 구현 기술에 의존되고 바뀔 일이 많은 RepositoryConfig는 리팩토링대로 분리해서 의존성을 파악하기 쉽게 하는 것이 좋겠습니다. 

그리고 Manager(FacadeConfig)는 따로 config로 관리할 필요성이 적고, 모듈의 근본적인 수정이 아닌 컨벤션 적용으로는 말씀해주신 스프링 의존성 문제를 해결하지 못하며 포이즈닝이 일어날 수 있으므로 리팩토링 전처럼 @Service를 적용하는 걸로 하면 좋겠습니다. :)

2eve2 commented on user-api/user-infrastructure/src/main/java/com/flab/user/infrastructure/config/UserRepositoryConfig.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class UserRepositoryConfig {
+
+    @Bean
+    public UserRepository userRepository(JpaUserRepository jpaUserRepository) {
+        return new UserRepositoryAdapter(jpaUserRepository);
+    }
+}
```
현재 Repository도 Bean으로 등록하셨는데, 코드 확인해보니 UserRepositoryAdapter에서 @Repository 어노테이션 제거 안하셨습니다. 혹시 확인해보셨을까요?

그리고 위와 같은 이유로 @Repository도 따로 Bean으로 등록해서 써야할 이유를 잘 모르겠습니다..!

dding94 replied:
해당 도메인은 제가 놓쳐서 작업을 못했네요 위에 얘기 나누고 방향을 정해서 고치면 될 것 같습니다.

2eve2 reviewed (approved):
논의한대로 @Service 부분만 다시 수정하시고 병합하면 될 것 같습니다.

dding94 committed:
refactor: 리뷰 반영

service 는 config 로 관리하는 것이 아닌 애플리케이션에서 사용하도록
gradle 의존성 트랜잭션과 component 를 사용하기 위해 둘다 포함된 spring data jpa 의존성을 사용
